contracts/dao/neutron-flashloans/src/contract.rs

```diff
@@ -0,0 +1,403 @@
+    // and the fee has been paid as (requested amount + fee).
+    let expected_balances = calculate_expected_balances(pre_loan_balances, fee.clone())?;
+
+    // Save all the information necessary to continue processing the loan request in the reply()
+    // handler.
+    ACTIVE_LOAN.save(
+        deps.storage,
+        &Some(ActiveLoan {
+            borrower: info.sender.clone(),
+            amount: amount.clone(),
+            fee,
+            expected_balances,
+        }),
+    )?;
+
+    // Send a (stargate -> authz -> bank) /cosmos.bank.v1beta1.MsgSend submessage with
+    // reply_on_success strategy (we want the transaction to be simply reverted in case of an
+    // error).
+    let msg_send = get_stargate_authz_bank_send_msg(env, config, info.sender, amount);
+    Ok(Response::new()
+        .add_submessage(SubMsg::reply_on_success(msg_send, AUTHZ_BANK_SEND_REPLY_ID))
+        .add_attribute("action", "execute_get_loan"))
+}
+
+// Check that the amount does not have duplicate or zero coins
+fn validate_amount(amount: Vec<Coin>) -> Result<(), ContractError> {
+    let mut denoms: HashSet<String> = HashSet::new();
+    for coin in amount {
+        if coin.amount.eq(&Uint128::zero()) {
+            return Err(ContractError::ZeroRequested { denom: coin.denom });
+        }
+
+        if denoms.contains(&coin.denom) {
+            return Err(ContractError::DuplicateDenoms {});
+        }
+
+        denoms.insert(coin.denom);
+    }
+
+    Ok(())
+}
+
+/// Returns the list of current balances on the contract's account for the coins that were
+/// requested by the borrower. If any actual coin balance is lower than the requested amount,
+/// returns an error.
+fn get_pre_loan_balances(
+    deps: &DepsMut,
+    source: Addr,
+    requested_amount: Vec<Coin>,
+) -> Result<Vec<Coin>, ContractError> {
+    // Prepare the query
+    let all_balances_query = BankQuery::AllBalances {
+        address: source.to_string(),
+    };
+    // Get the response (all balances)
+    let all_balances_response: AllBalanceResponse =
+        deps.querier.query(&all_balances_query.into())?;
+
+    // Filter all balances leaving only the balances of the requested coins
+    let mut pre_loan_balances: Vec<Coin> = vec![];
+    for requested_coin in requested_amount {
+        // Look for the requested coin in the source balances, AND check that the source
+        // has enough of the requested coin.
+        let maybe_source_coin = all_balances_response
+            .amount
+            .iter()
+            .find(|x| x.denom == requested_coin.denom && requested_coin.amount.le(&x.amount));
+
+        // If the source doesn't have (enough of) the requested coin, return an error
+        if maybe_source_coin.is_none() {
+            return Err(ContractError::InsufficientFunds {
+                denom: requested_coin.denom,
+            });
+        }
+
+        let source_coin = maybe_source_coin.unwrap();
+        pre_loan_balances.push(source_coin.clone())
```

Comment:
this can be rewritten in a cleaner way and without upwrap usage
```suggestion
        match maybe_source_coin {
            Some(coin) => pre_loan_balances.push(coin.clone()),
            // If the source doesn't have (enough of) the requested coin, return an error
            None => {
                return Err(ContractError::InsufficientFunds {
                    denom: requested_coin.denom,
                })
            }
        }
```